### PR TITLE
Global Illumination to WebGPU

### PR DESCRIPTION
**Pull request by deltakosh**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Review by sebavan — APPROVED** [review]

File: packages/dev/core/src/Rendering/GlobalIllumination/giRSMManager.ts [hunk 1]
```diff
@@ -30,10 +30,7 @@ import { expandToProperty, serialize } from "core/Misc/decorators";
 import { MaterialDefines } from "core/Materials/materialDefines";
 import { RegisterClass } from "core/Misc/typeStore";
 
-import "../../Shaders/bilateralBlur.fragment";
-import "../../Shaders/bilateralBlurQuality.fragment";
-import "../../Shaders/rsmGlobalIllumination.fragment";
-import "../../Shaders/rsmFullGlobalIllumination.fragment";
+import { ShaderLanguage } from "core/Materials";
```
Comment (noname0310) [inline packages/dev/core/src/Rendering/GlobalIllumination/giRSMManager.ts:33]:
```suggestion
import { ShaderLanguage } from "core/Materials/shaderLanguage";
```

File: packages/dev/core/src/Rendering/reflectiveShadowMap.ts [hunk 2]
```diff
@@ -433,19 +444,59 @@ export class RSMCreatePluginMaterial extends MaterialPluginBase {
...
+                #endif
+            `,
+
+                // eslint-disable-next-line @typescript-eslint/naming-convention
+                CUSTOM_FRAGMENT_BEFORE_FRAGCOLOR: `
+                #ifdef RSMCREATE
+                    var rsmColor = ${this._varAlbedoName} * uniforms.rsmLightColor;
+                    #ifdef RSMCREATE_PROJTEXTURE
+                    {
+                        var strq = rsmTextureProjectionMatrix * vec4f(fragmentInputs.vPositionW, 1.0);
```
Comment (noname0310) [inline packages/dev/core/src/Rendering/reflectiveShadowMap.ts:469]:
```suggestion
                        var strq = uniforms.rsmTextureProjectionMatrix * vec4f(fragmentInputs.vPositionW, 1.0);
```

File: packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    return clamp(indirectDiffuse * intensity, 0.0, 1.0);
+}
+
+
+@fragment
+fn main(input: FragmentInputs) -> FragmentOutputs {
+    var positionW: vec3f = textureSample(textureSampler, textureSamplerSampler, vUV).xyz;
+    var normalW: vec3f = textureSample(normalSampler, normalSamplerSampler, vUV).xyz;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx:61]:
```suggestion
    var positionW: vec3f = textureSample(textureSampler, textureSamplerSampler, fragmentInputs.vUV).xyz;
    var normalW: vec3f = textureSample(normalSampler, normalSamplerSampler, fragmentInputs.vUV).xyz;

```

File: packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    #ifdef DECODE_NORMAL
+        normalW = normalW * 2.0 - 1.0;
+    #endif
+    #ifdef TRANSFORM_NORMAL
+        normalW = (invView *  vec4f(normalW, 0.)).xyz;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx:66]:
```suggestion
        normalW = (uniforms.invView *  vec4f(normalW, 0.)).xyz;

```

File: packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    #endif
+
+    fragmentOutputs.color.rgb = computeIndirect(positionW, normalW);
+    fragmentOutputs.color.a = 1.0;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/rsmFullGlobalIllumination.fragment.fx:70]:
```suggestion
    fragmentOutputs.color = vec4f(computeIndirect(positionW, normalW), 1.0);
```

File: packages/dev/core/src/ShadersWGSL/rsmGlobalIllumination.fragment.fx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+@fragment
+fn main(input: FragmentInputs) -> FragmentOutputs {
+
+    var positionW: vec3f = textureSample(textureSampler, textureSamplerSampler, input.vUV).xyz;
+    var normalW: vec3f = textureSample(normalSampler, normalSamplerSampler, input.vUV).xyz;
+    #ifdef DECODE_NORMAL
+        normalW = normalW * 2.0 - 1.0;
+    #endif
+    #ifdef TRANSFORM_NORMAL
+        normalW = (invView *  vec4f(normalW, 0.)).xyz;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/rsmGlobalIllumination.fragment.fx:117]:
```suggestion
        normalW = (uniforms.invView *  vec4f(normalW, 0.)).xyz;
```

File: packages/dev/core/src/Rendering/GlobalIllumination/giRSMManager.ts [hunk 7]
```diff
@@ -283,6 +280,16 @@ export class GIRSMManager {
         this.recreateResources();
     }
 
+    /** Shader language used by the material */
+    protected _shaderLanguage = ShaderLanguage.GLSL;
+
+    /**
+     * Gets the shader language used in this material.
+     */
```
Comment (RaananW) [inline packages/dev/core/src/Rendering/GlobalIllumination/giRSMManager.ts:288]:
```suggestion
    /** Gets the shader language used in this material. */
```

(Force rebuild)

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15411/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15411/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15411/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15411/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15411/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15411/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15411/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15411/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15411/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15411/merge/testResults/webgl2playwright/index.html

**Pull request merged**